### PR TITLE
🔭 Nested scopes

### PR DESCRIPTION
**Pull request by favonia**
Let's have sections and their friends.

# Sections
```
section x begin
  def a : nat := 0
end
// x::a is visible
```

```
section x::y begin
  def a : nat := 0
end
// x::y::a is visible
```

```
section begin
  def a : nat := 0
end
// a is visible
```

```
def a : nat := 10
section begin
  def a : nat := 0
end
// errors due to unexpected shadowing
```

```
def a : nat := 10
!section begin
  def a : nat := 0
end
// do the shadowing (notice the bang)
```

```
section x begin
  def a : nat := 0
end [ a -> b ]
// x::b is visible
```

# New features of `import`

Now one can use `!import` instead of `import` to suppress the errors about shadowing.

# Remap what's visible

```
view [ a -> b ]
// rename a to b in what's currently visible (which does not affect what's being exported)

view [ m -> ]
// open everything under m

!view [ m -> ]
// open everything under m; shadow existing bindings with the bang
```

# Export new things from what's visible
```
export a
// export a from what's currently visible

export { a, b, c }
// export three bindings

export [ a; a -> b ]
// export a, but under the name b

export []
// export EVERYTHING that is visible (you probably do not want this)

!export { a, b, c }
// export a, b, and c; shadow existing a, b, c if necessary
```

# Remap what's being exported

```
repack [ a -> b ]
// manipulate what's being exported (which does not affect what's visible now)

!repack [ a -> b ]
// the variant that shadows existing bindings
```

# How to mimic X in Standard ML

Notice that untyped and ungrouped namespaces (here) are different from typed modules in Standard ML or OCaml.

## Open

```sml
open M
```
is
```
!view [ M -> ]
```

## Include

```sml
include M
```
is
```
!export [ M; M -> ]
!view [ M -> ]
```

## Rebind

```sml
module M = N
```
is
```
!view { N, [! M; N -> M] }
```

### CODE REVIEW

**Comment by favonia** [comment]
Blocked by #314.

**Comment by favonia** [comment]
Some ugliness: There are currently a "global" cofibration theory in `RefineState` and a "local" one in `RefineEnv`. `RefineEnv.append_con` and `RefineMonad.abstract` will update the local theory, while `RefineMonad.add_global` will update the global theory. When running NbE-related algorithms, we will use the conjunction of both cofibration theories. I am slightly worried about the inefficiencies due to the meet operation being called all the times, but feel the potential "more" efficient algorithm would require unnatural code structure.

**Review by jonsterling — APPROVED** [review]
This is really cool. I have not reviewed the code, I leave that to others, but the feature seems nicely designed. My one request is that the excellent summary that is contained in the PR text actually be brought into the repository, perhaps under the test/example.

File: test/section.cooltt
Comment (jonsterling) [inline test/section.cooltt:3]: Can you remind me what the question mark means here? I didn't see an example of it in the PR text.

Comment (favonia) [reply test/section.cooltt:3]: The question mark would show the current namespace. An example is in `import.cooltt`. `view` works on what's visible, so `view [ ?view ]` applies the pattern `?view` on what's visible, which ends up printing out what's visible without making any changes.

Comment (favonia) [reply test/section.cooltt:3]: Similarly, `repack [?label]` would inspect what's being exported and then leave it unchanged.

**Comment by favonia** [comment]
Another thing is that, while the current `def` would de the shadowing as before, we could change that and introduce `!def` for consistency. It means everything without the bang (`section`, `def`, `export`, `repack`, `view`) would err if the shadowing is happening, and everything with the bang (`!section`, `!def`, `!export`, `!repack`, `!view`) would happily shadow things.

**Comment by jonsterling** [comment]
@favonia That's a nice idea!

File: test/section.cooltt
Comment (jonsterling) [reply test/section.cooltt:3]: Ah ok. I think part of what confused me too was that I wasn't sure if the use of the identifier `?view` was meaningfully different from `?foo` in the example. Maybe the example can be changed to show that it doesn't matter what you put there, but I obvioulsy don't feel strongly about it.

This PR looks great!

**Comment by favonia** [comment]
I discussed this PR with @TOTBWF and decided to merge it first.

**Pull request merged**
